tests/blockchain/eth/contracts/main/user_escrow/test_user_escrow_proxy.py

```diff
@@ -0,0 +1,250 @@
+        tx = linker_contract.functions.upgrade(library_v2.address, secret2, secret2_hash).transact({'from': creator})
+        testerchain.wait_for_receipt(tx)
+
+    # Creator can't use the same secret again because it's insecure
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = linker_contract.functions.upgrade(library_v2.address, secret, secret_hash).transact({'from': creator})
+        testerchain.wait_for_receipt(tx)
+
+    assert library_v1.address == linker_contract.functions.target().call()
+    tx = linker_contract.functions.upgrade(library_v2.address, secret, secret2_hash).transact({'from': creator})
+    testerchain.wait_for_receipt(tx)
+    assert library_v2.address == linker_contract.functions.target().call()
+
+    # Method with old signature is not working
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = user_escrow_library_v1.functions.firstMethod().transact({'from': user})
+        testerchain.wait_for_receipt(tx)
+    # Method with old signature that available in new ABI is working
+    assert 15 == user_escrow_library_v1.functions.secondMethod().call({'from': user})
+
+    # New ABI is working
+    assert 15 == user_escrow_library_v2.functions.secondMethod().call({'from': user})
+    tx = user_escrow_library_v2.functions.firstMethod(10).transact({'from': user})
+    testerchain.wait_for_receipt(tx)
+    tx = user_escrow_library_v2.functions.thirdMethod().transact({'from': user})
+    testerchain.wait_for_receipt(tx)
+
+    # And can send and withdraw ETH
+    tx = testerchain.interface.w3.eth.sendTransaction(
+        {'from': user, 'to': user_escrow_contract.address, 'value': 1, 'gas_price': 0})
+    testerchain.wait_for_receipt(tx)
+    assert 1 == testerchain.interface.w3.eth.getBalance(user_escrow_contract.address)
+    # Only user can send ETH
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = testerchain.interface.w3.eth.sendTransaction(
+            {'from': testerchain.interface.w3.eth.coinbase,
+             'to': user_escrow_contract.address,
+             'value': 1,
+             'gas_price': 0})
+        testerchain.wait_for_receipt(tx)
+    assert 1 == testerchain.interface.w3.eth.getBalance(user_escrow_contract.address)
+
+    rewards = user_escrow_contract.events.ETHWithdrawn.createFilter(fromBlock='latest')
+    user_balance = testerchain.interface.w3.eth.getBalance(user)
+    tx = user_escrow_contract.functions.withdrawETH().transact({'from': user, 'gas_price': 0})
+    testerchain.wait_for_receipt(tx)
+    assert user_balance + 1 == testerchain.interface.w3.eth.getBalance(user)
+    assert 0 == testerchain.interface.w3.eth.getBalance(user_escrow_contract.address)
+
+    events = rewards.get_all_entries()
+    assert 1 == len(events)
+    event_args = events[0]['args']
+    assert user == event_args['owner']
+    assert 1 == event_args['value']
+
+
+@pytest.mark.slow
+def test_proxy_selfdestruct(testerchain, token):
```

Comment:
Perhaps I missed this - Have we introduced a self destructible proxy for production use?

Reply:
Not yet, if some day it will be or will be possible to destroy contract even without special method - then `Dispatcher` and `UserEscrow` must work still correctly

Reply:
:ok_hand: